### PR TITLE
[RNG] Updated Device API

### PR DESCRIPTION
**Pull request by andreyfe1**
# Description

Added following distributions:
* `beta`
* `gamma`.

Added support of `int64_t`, `uint64_t` for `uniform`.
Added support of `int8_t`, `uint8_t`, `int16_t`, `uint16_t` for `bernoulli`.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
- [x] Have you formatted the code using clang-format?

## New interfaces

- [x] What version of oneAPI spec the interface is targeted?
       [v1.4-provisional-rev-1](https://github.com/uxlfoundation/oneAPI-spec/releases/tag/oneAPI-v1.4-provisional-rev-1)
- [ ] Complete [New features](pull_request_template.md#new-features) checklist

## New features

- [X] Have you provided motivation for adding a new feature?
- [X] Have you added relevant tests?

### CODE REVIEW

File: include/oneapi/mkl/rng/device/detail/mcg31m1_impl.hpp [hunk 1]
```diff
@@ -56,7 +56,7 @@ constexpr sycl::vec<std::uint64_t, VecSize> select_vector_a_mcg31m1() {
 // hipSYCL (AdaptiveCpp) doesn't support constexpr sycl::vec constructor
 // that's why in case of hipSYCL backend sycl::vec is created as a local variable
 #ifndef __HIPSYCL__
-template <std::int32_t VecSize>
+template <std::uint64_t VecSize>
```
Comment (iMartyan) [inline include/oneapi/mkl/rng/device/detail/mcg31m1_impl.hpp:59]: Could you please explain why change length of vecsize?

File: include/oneapi/mkl/rng/device/detail/uniform_impl.hpp [hunk 2]
```diff
@@ -73,25 +121,97 @@ class distribution_base<oneapi::mkl::rng::device::uniform<Type, Method>> {
...
+                    return res;
+                }
             }
         }
         else {
             res = engine.generate(a_, b_);
             if constexpr (std::is_same<Method, uniform_method::accurate>::value) {
-                res = sycl::fmax(res, OutType{ a_ });
-                res = sycl::fmin(res, OutType{ b_ });
+                res = std::fmax(res, a_);
```
Comment (iMartyan) [inline include/oneapi/mkl/rng/device/detail/uniform_impl.hpp]: why `std` func?

File: include/oneapi/mkl/rng/distributions.hpp [hunk 3]
```diff
@@ -61,17 +61,24 @@ template <typename Type = float, typename Method = uniform_method::by_default>
...
+                      std::is_same<Type, std::uint32_t>::value,
                   "rng uniform distribution type is not supported");
 
     using method_type = Method;
     using result_type = Type;
 
-    uniform() : uniform(static_cast<Type>(0.0f), static_cast<Type>(1.0f)) {}
+    uniform()
+            : uniform(static_cast<Type>(0.0f),
+                      std::is_integral<Type>::value
```
Comment (iMartyan) [inline include/oneapi/mkl/rng/distributions.hpp]: Do we need to add similar `if` for 64 bit integers here similar to device/distributions/hpp ?

**Review by iMartyan — COMMENTED** [review]
Please also expand tests for new types for uniform and Bernoulli distributions

File: include/oneapi/mkl/rng/device/detail/mcg31m1_impl.hpp (hunk 1 above)
Comment (iMartyan) [reply include/oneapi/mkl/rng/device/detail/mcg31m1_impl.hpp:59]: decided offline

File: include/oneapi/mkl/rng/distributions.hpp (hunk 3 above)
Comment (andreyfe1) [reply include/oneapi/mkl/rng/distributions.hpp]: Reverting changes for host API. Let's concentrate on device API only

**Review by iMartyan — APPROVED** [review]
lgtm
Thanks!

File: tests/unit_tests/rng/device/include/rng_device_test_common.hpp [hunk 4]
```diff
@@ -226,6 +226,44 @@ struct statistics_device<oneapi::mkl::rng::device::uniform<std::uint32_t, Method
...
+             (7.0 / 240.0);
+
+        return compare_moments(r, tM, tD, tQ);
+    }
+};
+
+template <typename Method>
+struct statistics_device<oneapi::mkl::rng::device::uniform<std::uint64_t, Method>> {
+    template <typename AllocType>
+    bool check(const std::vector<uint64_t, AllocType>& r,
```
Comment (aelizaro) [inline tests/unit_tests/rng/device/include/rng_device_test_common.hpp]:
```suggestion
    bool check(const std::vector<std::uint64_t, AllocType>& r,
```
std:: is missing in some places here

File: tests/unit_tests/rng/device/moments/moments.cpp [hunk 5]
```diff
@@ -961,6 +1049,190 @@ INSTANTIATE_TEST_SUITE_P(Philox4x32x10ExponentialIcdfAccDeviceMomentsTestsSuite,
...
+        : public ::testing::TestWithParam<sycl::device*> {};
+
+class Philox4x32x10BetaCjaAccDeviceMomentsTests
+        : public ::testing::TestWithParam<sycl::device*> {};
+
+// implementation uses double precision for accuracy
+TEST_P(Philox4x32x10BetaCjaDeviceMomentsTests, RealDoublePrecision) {
+    CHECK_DOUBLE_ON_DEVICE(GetParam());
+
+    rng_device_test<moments_test<oneapi::mkl::rng::device::philox4x32x10<1>,
```
Comment (aelizaro) [inline tests/unit_tests/rng/device/moments/moments.cpp]: I propose to split float and double, if we can still test float on HW without FP64 support

**Review by aelizaro — APPROVED** [review]
Looks good to me, thank you!

File: include/oneapi/mkl/rng/device/detail/uniform_impl.hpp (hunk 2 above)
Comment (Rbiessy) [reply include/oneapi/mkl/rng/device/detail/uniform_impl.hpp]: I am also curious about this question. Could you share what the reason was?

Comment (andreyfe1) [reply include/oneapi/mkl/rng/device/detail/uniform_impl.hpp]: It was an intermediate solution. However, if we use the separate branch for HIP SYCL it's fine to keep `sycl`

Comment (Rbiessy) [reply include/oneapi/mkl/rng/device/detail/uniform_impl.hpp]: Ok thanks. There has been some issues with using `std` functions in SYCL kernels before so it's good that you reverted this IMO.

Comment (andreyfe1) [reply include/oneapi/mkl/rng/device/detail/uniform_impl.hpp]: The funny thing is that I experienced issues with `sycl::fmax` using `icpx` that's why I changed `sycl` to `std` :)

Comment (Rbiessy) [reply include/oneapi/mkl/rng/device/detail/uniform_impl.hpp]: That is concerning if that's the case!

**Pull request merged**